keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -348,25 +349,117 @@ to fail with a descriptive error.
+- 
+- `charAt`
+- `indexOf`, `lastIndexOf`
+- `upperAscii`, `lowerAscii`
+- `replace` (string literal replacement, not regex)
+- `split`
+- `substring`
+- `trim`
+
+Notable absences from the standard functions, when comparing against the core libraries of Go, Java and Python are:
+
+Strings:
+
+- trim/trimLeft/trimRight (trimming either by whitespace or by provided cutset)
+- trimPrefix/trimSuffix
+- join
+- replace (regex)
+- format (e.g. `fmt.Sprintf()`)
+
+Lists:
+
+- sort
+- indexOf/lastIndexOf
+- replace
+- sublist
+- split
+- reverse
+- sum/min/max (xref: [Rego function library](https://www.openpolicyagent.org/docs/latest/policy-reference/#built-in-functions))
+
+Numbers:
+
+- abs/cel/floor/round
+- exp/log/log10/logN/pow/sqrt
+- max(x, y, ..)/min(x, y, ...)
+- <trig functions>
+
+Formats:
 
+- semver: compare (xref [Kyverno function library](https://github.com/kyverno/kyverno/blob/main/pkg/engine/jmespath/functions.go))
+
+
+Maps:
+
+- <Ability to construct a map using a comprehension>
+
+Some considerations when selecting which of the above we should include in CEL expressions in Kubernetes:
+
+- We should include functions that will be needed for future use cases of CEL in Kubernetes (e.g. general admission) as
+  ones needed for validation.
+- Regex replace is very powerful and useful. It is also potentially dangerous due to its ability to allocate memory.
+- `format`: Since CEL supports string concatenation, the value of having format would only be to do things like format floats.
+  Formatting functions are complex and require extensive documentation to teach.
+- `isSorted` makes it possible to check if a list is sorted without the expensive of performing a sort
+- `indexOf` / `lastIndexOf` / `split` / `replace` (the string functions) will be overloaded to provide the equivalent list functions.
+- `reduce` is going to be non-obvious to anyone that doesn't have a functional programming background?
+  - If we provide `reduce` are developers going to also expect to have `fold` or `zip`?
+- "Ability to construct a map using a comprehension": this appears mechanically problematic to support in CEL?
+- None of the other policy libraries provide extended math/trig support and I asked Tristan (maintainer of CEL) if
+  they are requested and he said "almost never", which he clarified to mean it has been requested exactly one time.
+- Average and quantile (median, 99th percentile, ...) and stddev are, however, often requested by other CEL users
+- `hasPrefix` / `hasSuffix` is useful but an be performed trivially using regex matching
+
+Proposal:
+
+- Keep all the CEL standard functions and macros as well as the extended string library.
+- Add `isSorted` for lists with comparable elements. This is useful for ensuring that a list is kept in-order.
+- Add `sum`, `min` and `max` functions for lists of summable/comparable elements. This is the
```

Comment:
what is the overflow behavior for sum? (also, is there defined overflow behavior for addition operations in cel)

Reply:
clarify what summable/comparable elements are?

Reply:
this comment is still outstanding

Reply:
Fixed. I've referenced https://github.com/google/cel-spec/blob/master/doc/langdef.md#overflow and the + and comparison operators listed in https://github.com/google/cel-spec/blob/master/doc/langdef.md#list-of-standard-definitions and explained the behavior.